File: graph-node/graph.js
// class Graph {
//   constructor() {
//     this._edges = {}
//     this._inNeighbors = {}
//   }

//   connect(vertexFrom, vertexTo, weight) {
//     vertexFrom = vertexFrom.toString()
//     vertexTo = vertexTo.toString()

//     if (!this._edges.hasOwnProperty(vertexFrom)) {
//       this._edges[vertexFrom] = {}
//       this._inNeighbors[vertexFrom] = new Set()
//     }

//     if (!this._edges.hasOwnProperty(vertexTo)) {
//       this._edges[vertexTo] = {}
//       this._inNeighbors[vertexTo] = new Set()
//     }

//     this._edges[vertexFrom][vertexTo] = weight
//     this._inNeighbors[vertexTo].add(vertexFrom)
//   }

//   findOutNeighbors(vertex) {
//     return new Set(Object.keys(this._edges[vertex]))
//   }

//   findInNeighbors(vertex) {
//     return new Set(this._inNeighbors[vertex])
//   }

//   findVertices() {
//     return new Set(Object.keys(this._edges))
//   }

//   findConnectedComponents() {
//     const unvisited = this.findVertices()

//     const L = []

//     for (const vertex of this.findVertices()) {
//       this.visit(vertex, unvisited, L)
//     }

//     const components = {}

//     for (let i = L.length - 1; i >= 0; i -= 1) {
//       this.assign(L[i], L[i], components)
//     }

//     return components
//   }

//   visit(vertex, unvisited, L) {
//     if (unvisited.has(vertex)) {
//       unvisited.delete(vertex)

//       for (const neighbor of this.findOutNeighbors(vertex)) {
//         this.visit(neighbor, unvisited, L)
//       }

//       L.push(vertex)
//     }
//   }

//   assign(vertex, root, components) {
//     let hasBeenAssigned = false

//     for (const component of Object.values(components)) {
//       if (component.has(vertex)) {
//         hasBeenAssigned = true
//         break
//       }
//     }

//     if (!hasBeenAssigned) {
//       if (!components.hasOwnProperty(root)) {
//         components[root] = new Set()
//       }

//       components[root].add(vertex)

//       for (const neighbor of this.findInNeighbors(vertex)) {
//         this.assign(neighbor, root, components)
//       }
//     }
//   }
// }

class Graph {
  constructor() {
    this._edges = new Map()
    this._inNeighbors = new Map()
  }

  connect(vertexFrom, vertexTo, weight) {
    if (!this._edges.has(vertexFrom)) {
      this._edges.set(vertexFrom, new Map())
      this._inNeighbors.set(vertexFrom, new Set())
    }

    if (!this._edges.has(vertexTo)) {
      this._edges.set(vertexTo, new Map())
      this._inNeighbors.set(vertexTo, new Set())
    }

    this._edges.get(vertexFrom).set(vertexTo, weight)
    this._inNeighbors.get(vertexTo).add(vertexFrom)
  }

  findOutNeighbors(vertex) {
    return new Set(this._edges.get(vertex).keys())
  }

  findInNeighbors(vertex) {
    return new Set(this._inNeighbors.get(vertex))
  }

  findVertices() {
    return new Set(this._edges.keys())
  }

  findConnectedComponents() {
    const unvisited = this.findVertices()

    const L = []

    for (const vertex of this.findVertices()) {
      this.visit(vertex, unvisited, L)
    }

    const components = new Map()

    for (let i = L.length - 1; i >= 0; i -= 1) {
      this.assign(L[i], L[i], components)
    }

    return components
  }

  visit(vertex, unvisited, L) {
    if (unvisited.has(vertex)) {
      unvisited.delete(vertex)

      for (const neighbor of this.findOutNeighbors(vertex)) {
        this.visit(neighbor, unvisited, L)
      }

      L.push(vertex)
    }
  }

  assign(vertex, root, components) {
    let hasBeenAssigned = false

    for (const component of components.values()) {
      if (component.has(vertex)) {
        hasBeenAssigned = true
        break
      }
    }

    if (!hasBeenAssigned) {
      if (!components.has(root)) {
        components.set(root, new Set())
      }

      components.get(root).add(vertex)

      for (const neighbor of this.findInNeighbors(vertex)) {
        this.assign(neighbor, root, components)
      }
    }
  }
}

module.exports = { Graph }

// const adjacenceMatrix = [
//     [0, 1, 0, 0, 0, 0, 0, 0],
//     [0, 0, 1, 1, 0, 0, 0, 0],
//     [1, 0, 0, 1, 0, 0, 0, 0],
//     [0, 0, 0, 0, 1, 0, 0, 0],
//     [0, 0, 0, 1, 0, 0, 0, 0],
//     [0, 0, 0, 0, 1, 0, 1, 0],
//     [0, 0, 0, 0, 0, 1, 0, 1],
//     [0, 0, 0, 0, 1, 0, 1, 0],
// ]

File: graph-python/graph.py
from typing import Dict, Any, Set, List
from collections import deque

class Graph():
    _edges: Dict[Any, Dict[Any, int]]
    _in_neighbors: Dict[Any, Set[Any]]

    def __init__(self):
        self._edges = {}
        self._in_neighbors = {}

    def connect(self, vertex_from: Any, vertex_to: Any, weight: int):
        if vertex_from not in self._edges:
            self._edges[vertex_from] = {}
            self._in_neighbors[vertex_from] = set()

        if vertex_to not in self._edges:
            self._edges[vertex_to] = {}
            self._in_neighbors[vertex_to] = set()

        self._edges[vertex_from][vertex_to] = weight
        self._in_neighbors[vertex_to].add(vertex_from)

    def find_all_out_neighbors(self, vertex):
        return set(self._edges[vertex].keys())

    def find_all_in_neighbors(self, vertex):
        return set(self._in_neighbors[vertex])

    def find_all_vertices(self):
        return set(self._edges.keys())

    def find_connected_components(self):
        unvisited = self.find_all_vertices()

        L = deque()

        for vertex in self.find_all_vertices():
            self._visit(vertex, unvisited, L)

        components: Dict[Any, Set[Any]] = {}

        for vertex in L:
            self._assign(vertex, vertex, components)

        return components

    def _visit(self, vertex: Any, unvisited: Set[Any], L: deque):
        if vertex in unvisited:
            unvisited.remove(vertex)

            for neighbor in self.find_all_out_neighbors(vertex):
                self._visit(neighbor, unvisited, L)

            L.appendleft(vertex)

    def _assign(self, vertex: Any, root: Any, components: Dict[Any, Set[Any]]):
        has_been_assigned = False

        for component in components.values():
            if vertex in component:
                has_been_assigned = True
                break

        if not has_been_assigned:
            if root not in components:
                components[root] = set()

            components[root].add(vertex)

            for neighbor in self.find_all_in_neighbors(vertex):
                self._assign(neighbor, root, components)

def main():
    graph = Graph()

    graph.connect('1', '2', 1)
    graph.connect('2', '3', 1)
    graph.connect('2', '4', 1)
    graph.connect('3', '1', 1)
    graph.connect('3', '4', 1)

    graph.connect('4', '5', 1)
    graph.connect('5', '4', 1)

    graph.connect('6', '7', 1)
    graph.connect('6', '5', 1)
    graph.connect('7', '6', 1)
    graph.connect('7', '8', 1)
    graph.connect('8', '7', 1)
    graph.connect('8', '5', 1)

    print(graph.find_connected_components())

if __name__ == '__main__':
    main()
